### PR TITLE
43 model c framework

### PR DESCRIPTION
**Pull request by wkCircle**
This PR is related to #43 

There are rooms of improvement but at least we have the following functionality with the PR:
- can record user voice (given only fixed duration of time in advance)
- use default speaker or the user voice to synthesize target sentence (Text to Speech)

### CODE REVIEW

File: experiments/model_c/api.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+
+def get_device() -> str:
+    """Check if cuda is available and return corresp. string"""
+    device = "cuda" if torch.cuda.is_available() else "cpu"
+    return device
+
+
+def get_tts_model(
+    identifier: Optional[str] = None, device: DeviceLikeType = "cpu"
```
Comment (johnklee) [inline experiments/model_c/api.py:22]: Input argument `device` is missed in docstr. FYI.

File: experiments/model_c/api.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    identifier: str
+        supported model name of format <model_category> / <language_code> /
+        <voice_or_dataset_category> / <model_name>. For more details, please
+        check with the code ``TTS().list_models().list_models()``.
+        Defaults to "tts_models/multilingual/multi-dataset/xtts_v2" if None.
+    """
+    if identifier is None:
+        # identifier = "tts_models/multilingual/multi-dataset/your_tts"
+        # Notice that XTTSv2 has license for non-commerical use only.
+        identifier = "tts_models/multilingual/multi-dataset/xtts_v2"
```
Comment (johnklee) [inline experiments/model_c/api.py:39]: Nit: FYI
```python
identifier = identifier or "tts_models/multilingual/multi-dataset/xtts_v2"
```

File: experiments/model_c/demo.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+          tts.synthesizer.tts_model.speaker_manager.name_to_id
+    )
+    print("list of available languages:\n", tts.languages)
+
+    # record voice (can be optional or conditional in the future)
+    print("Please record your voice for references")
+    _ = record_voice(duration=duration, file_path=speaker_wav)
+
+    # inference
+    # one of the spkear and speaker_wav (6-10 secs) must be given
```
Comment (johnklee) [inline experiments/model_c/demo.py:32]: typo? `speaker`

File: experiments/model_c/api.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+or inferencing. Here we deomnstrate how to use python to call models
+"""
+from typing import Optional, Union
+import sounddevice as sd
+from scipy.io import wavfile
+import torch
+from TTS.api import TTS
+
+
+DeviceLikeType = Union[str, torch.device, int]
```
Comment (johnklee) [inline experiments/model_c/api.py:12]: From 3.11, you could write it this way:
```python
from typing import TypeAlias

DeviceLikeType: TypeAlias = str | torch.device | int
```

Reference: https://peps.python.org/pep-0604/

**Review by johnklee — COMMENTED** [review]
Few options for your reference.

File: experiments/model_c/api.py (hunk 4 above)
Comment (wkCircle) [reply experiments/model_c/api.py:12]: updated, thanks!

File: experiments/model_c/api.py (hunk 1 above)
Comment (wkCircle) [reply experiments/model_c/api.py:21]: solved, thanks!

File: experiments/model_c/api.py (hunk 2 above)
Comment (wkCircle) [reply experiments/model_c/api.py]: solved, thanks!

**Review by johnklee — APPROVED** [review]
LGTM

**Comment by wkCircle** [comment]
@johnklee  I need your approval again cuz I just updated the PR to solve the conflict for 
- poetry.lock
- pyproject.toml 

for package dependencies. Thanks.

File: experiments/model_c/api.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+
+def get_device() -> str:
+    """Check if cuda is available and return corresp. string"""
+    device = "cuda" if torch.cuda.is_available() else "cpu"
+    return device
+
+
+def get_tts_model(
+    identifier: Optional[str] = None, device: DeviceLikeType = "cpu"
```
Comment (johnklee) [inline experiments/model_c/api.py:21]: Nit: `identifier: str | None = None`

File: experiments/model_c/demo.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    # inputs that can be put into argparser with proper defaults in the future
+    intput_text = "我想要吃飯，你有推薦的Restaurant嗎?"
+    speaker = "Luis Moray"
+    language = "zh-cn"
+    file_path = "./output.wav"
+    duration = 10
+    speaker_wav = "./experiments/model_c/voice_ref/chung_recorded.wav"
+
+    # retrieve model
+    device = get_device()
```
Comment (johnklee) [inline experiments/model_c/demo.py:18]: `device` is never used or referred the follow up. Do we need this line?

File: experiments/model_c/api.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+import torch
+from TTS.api import TTS
+
+DeviceLikeType: TypeAlias = str | torch.device | int
+
+
+def get_device() -> str:
+    """Check if cuda is available and return corresp. string"""
+    device = "cuda" if torch.cuda.is_available() else "cpu"
+    return device
```
Comment (johnklee) [inline experiments/model_c/api.py:17]: Nit: `return "cuda" if torch.cuda.is_available() else "cpu"`

**Review by johnklee — APPROVED** [review]
LGTM

File: experiments/model_c/api.py (hunk 5 above)
Comment (wkCircle) [reply experiments/model_c/api.py:21]: or can I type: `identifier: str = None`?

File: experiments/model_c/api.py (hunk 7 above)
Comment (wkCircle) [reply experiments/model_c/api.py:17]: fixed, thanks

File: experiments/model_c/demo.py (hunk 6 above)
Comment (wkCircle) [reply experiments/model_c/demo.py:18]: Actually, `get_tts_model` should use the device, I have merged the line as an argument of the `get_tts_model`, FYI.

File: experiments/model_c/api.py (hunk 5 above)
Comment (johnklee) [reply experiments/model_c/api.py]: `identifier: str` indicated that `identifier` is an instance of `str`. But `None` is not the type of `str`:
```python
>>> isinstance('str', str)
True
>>> isinstance(None, str)
False
```
`identifier: str = None` will fail in typing check from our internal build build system. Just for your reference and I am ok with your suggestion while we don't really apply for typing check currently. FYI.

File: experiments/model_c/demo.py (hunk 6 above)
Comment (johnklee) [reply experiments/model_c/demo.py]: Understood. Thanks for the clarification.

**Review by johnklee — APPROVED** [review]
Sorry for the late response. LGTM.

**Pull request merged**